crates/bevy_app/src/app.rs

```diff
@@ -762,11 +770,41 @@ impl App {
     where
         T: Plugin,
     {
-        debug!("added plugin: {}", plugin.name());
+        debug!("added plugin {}", plugin.name());
+        if plugin.is_unique() {
+            self.register_plugin(&std::any::TypeId::of::<T>(), plugin.name(), None);
+        }
         plugin.build(self);
         self
     }
 
+    /// Checks that a plugin has not already been added to an application. It will panic with an
+    /// helpful message the second time a plugin is being added.
+    pub(crate) fn register_plugin(
+        &mut self,
+        plugin_type: &TypeId,
+        plugin_name: &str,
+        from_group: Option<&'static str>,
+    ) {
+        if let Some(existing_from_group) = self.plugins.insert(*plugin_type, from_group) {
+            match (from_group, existing_from_group) {
+                (None, None) => panic!("Plugin \"{}\" was already added", plugin_name),
```

Comment:
Before we add panics here, I think we should discuss moving in the direction of "plugins add the plugins they depend on and bevy only initializes each plugin once". With this, dependencies would need to be implicit. But we could make them explicit if we just de-dupe unique plugins.

However this gets a bit messy if we couple configuration to specific plugin instances, because then we've added the ability to init with multiple conflicting configs.

Reply:
I agree, a more comprehensive solution with some more involved architectural design of how plugins work would be my preferred solution. You’re small but significant fix of collecting all the plugins and deduping then initializing is good and easier fix. At some point bevy will need a more sophisticated plugin architecture, do you have an idea when you would want to work on that (or just let alice do it in her existing rfc for this more complex but more robust plugin architecture after stageless is rolled out?